### ISSUE TITLE
[BUG]: RSS Guard gets confused with entries missing an <author> tag

### ISSUE THREAD

**Issue by guihkx** [martinrotter/rssguard#412]
#### Brief description of the issue.
I'm having a weird issue with [this Reddit feed](https://old.reddit.com/r/rust/new/.rss) right now.

Basically, every time there's a new entry, RSS Guard creates a duplicate of an old entry, but it changes the authorship of the entry, adding all previous authors in the whole feed (lol):

![image](https://user-images.githubusercontent.com/626206/118416615-f65b8d00-b686-11eb-8bf0-b0abd35febf4.png)

#### How to reproduce the bug?
I'm not entirely sure, but I suspect the issue is triggered when an user posts something on Reddit and delete their account. After that, Reddit unsets the `<author>` tag and then maybe RSS Guard can't handle that.

At least that's the only different thing I could noticed about this duplicate entry, is that it doesn't have an `<author>` tag (because the user deleted his account, most likely), e.g.:

```xml
<entry>
    <category term="rust" label="r/rust"/>
    <content type="html">&lt;br/&gt; &lt;span&gt;&lt;a href=&quot;https://github.com/akhramov/knast&quot;&gt;[link]&lt;/a&gt;&lt;/span&gt; &amp;#32; &lt;span&gt;&lt;a href=&quot;https://old.reddit.com/r/rust/comments/nd5enm/knast_freebsd_oci_container_runtime_written_in/&quot;&gt;[comments]&lt;/a&gt;&lt;/span&gt;</content>
    <id>t3_nd5enm</id>
    <link href="https://old.reddit.com/r/rust/comments/nd5enm/knast_freebsd_oci_container_runtime_written_in/" />
    <updated>2021-05-15T18:34:53+00:00</updated>
    <published>2021-05-15T18:34:53+00:00</published>
    <title>KNAST - FreeBSD OCI container runtime written in Rust</title>
</entry>
```

While all the other entries do have the `<author>` tag, e.g.:

```xml
<entry>
    <author>
        <name>/u/habitue</name>
        <uri>https://old.reddit.com/user/habitue</uri>
    </author>
    <category term="rust" label="r/rust"/>
    <content type="html">&amp;#32; submitted by &amp;#32; &lt;a href=&quot;https://old.reddit.com/user/habitue&quot;&gt; /u/habitue &lt;/a&gt; &lt;br/&gt; &lt;span&gt;&lt;a href=&quot;https://blog.abor.dev/p/timclicks?r=jx6d6&amp;amp;utm_campaign=post&amp;amp;utm_medium=web&amp;amp;utm_source=reddit&quot;&gt;[link]&lt;/a&gt;&lt;/span&gt; &amp;#32; &lt;span&gt;&lt;a href=&quot;https://old.reddit.com/r/rust/comments/nd646r/routes_to_discovering_rust/&quot;&gt;[comments]&lt;/a&gt;&lt;/span&gt;</content>
    <id>t3_nd646r</id>
    <link href="https://old.reddit.com/r/rust/comments/nd646r/routes_to_discovering_rust/" />
    <updated>2021-05-15T19:08:17+00:00</updated>
    <published>2021-05-15T19:08:17+00:00</published>
    <title>Routes to Discovering Rust</title>
</entry>
```

#### What was the expected result?
RSS Guard shouldn't duplicate entries without an `<author>` tag (?).

#### What actually happened?
RSS Guard duplicates weird entries, even though I'm using a message filter that explicitly ignores new entries if their URL already exists in RSS Guard's database:

```js
function filterMessage() {
  if (msg.isDuplicateWithAttribute(MessageObject.SameUrl | MessageObject.AllFeedsSameAccount)) {
    return MessageObject.Ignore;
  }
  return MessageObject.Accept;
}
```

#### Other information
The buggy entry will probably go away very soon, so it will be harder to reproduce this, so here's a backup of the feed:

[reddit_rust_feed.xml.zip](https://github.com/martinrotter/rssguard/files/6490262/reddit_rust_feed.xml.zip)

Then perhaps we have an easier way to reproduce it, I don't know. :p 

* OS: Arch Linux
* RSS Guard version: 3.9.2

**Comment by guihkx** [comment]
![image](https://user-images.githubusercontent.com/626206/118420222-b7cccf00-b694-11eb-98e9-46445ab5326d.png)

**Comment by martinrotter** [comment]
> 
> 
> ![image](https://user-images.githubusercontent.com/626206/118420222-b7cccf00-b694-11eb-98e9-46445ab5326d.png)

I am confused by this picture. Why are those two articles highlighted? They have completely different titles...

**Comment by martinrotter** [comment]
As for "authors" I fixed it right now and indeed there was an error, quite bad one.

As for "even though I'm using a message filter that explicitly ignores new entries if their URL" if you feel that message filtering does not work properly then file separate bug ticket and provide application debug log :)

**Comment by guihkx** [comment]
Thanks for fixing it! :D 

>I am confused by this picture. Why are those two articles highlighted? They have completely different titles...

I was highlighting the author's name, actually.

>As for "even though I'm using a message filter that explicitly ignores new entries if their URL" if you feel that message filtering does not work properly then file separate bug ticket and provide application debug log :)

I'll see if I can reproduce it now. Thanks

**Comment by martinrotter** [comment]
@guihkx OK, it is fair to say that the actual RSS/RDF/ATOM parsing code is not of best quality, I should rewrite it someday to make it easier to maintain.

**Comment by guihkx** [comment]
That's okay :)

By the way, just tested the changes and it completely fixed it, so thank you. The issue I mentioned about the message filter not ignoring duplicate URLs seems to be gone as well. 🎉 

Though I have a different issue regarding the message filters that I'll report soon.

**Comment by guihkx** [comment]
You know what, come to think about it, I don't think it's a bug...

You see, I have this message filter to ignore duplicate URLs:

```js
function filterMessage() {
  if (msg.isDuplicateWithAttribute(MessageObject.SameUrl | MessageObject.AllFeedsSameAccount)) {
    return MessageObject.Ignore;
  }
  return MessageObject.Accept;
}
```

Now assume I have an entry (or message, if you will) on a random feed that points to this URL: `https://github.com/martinrotter/rssguard/issues/412`

If I delete that entry (including from the recycle bin), RSS Guard will not re-add the message when I update that feed. Which seems the correct behavior, IMO.

So in short: I thought there was a bug in the message filter, because when I deleted an entry (URL: `https://github.com/martinrotter/rssguard/issues/412`) and tried to test the filter using that same URL, RSS Guard would _not_ accept the message, which at first glance didn't make sense because I had deleted it.

**Comment by martinrotter** [comment]
> If I delete that entry (including from the recycle bin), RSS Guard will not re-add the message when I update that feed. Which seems the correct behavior, IMO.

Yes, if you "delete" the entry, then it is still kept in RSS Guard's database, it is just marked as "deleted" (or "permanently deleted" if deleting from bin) precisely for these situations. There are some corner cases where either behavior is "weird" thus there is no single behavior which would satisfy everyone. Over the years I figured out that keeping "deleted" articles and including them in RSS Guard internal logic is best.

I could probably add `MessageObject.OnlyUndeletedMessages` flag to `isDuplicateWithAttribute(...)` function to allow users to tweak behavior of the test.

**Comment by guihkx** [comment]
>Over the years I figured out that keeping "deleted" articles and including them in RSS Guard internal logic is best.

Agreed :)

> I could probably add `MessageObject.OnlyUndeletedMessages` flag to `isDuplicateWithAttribute(...)` function to allow users to tweak behavior of the test.

I'm never against new features, so... 😄
